src/http/one/TeChunkedParser.cc

```diff
@@ -115,55 +116,60 @@ Http::One::TeChunkedParser::parseChunkSize(Http1::Tokenizer &tok)
  * ICAP 'use-original-body=N' extension is supported.
  */
 bool
-Http::One::TeChunkedParser::parseChunkExtension(Http1::Tokenizer &tok, bool skipKnown)
+Http::One::TeChunkedParser::parseChunkExtensions(Tokenizer &tok, const bool skipKnown)
 {
-    SBuf ext;
-    SBuf value;
-    while (
-        ParseBws(tok) && // Bug 4492: IBM_HTTP_Server sends SP after chunk-size
-        tok.skip(';') &&
-        ParseBws(tok) && // Bug 4492: ICAP servers send SP before chunk-ext-name
-        tok.prefix(ext, CharacterSet::TCHAR)) { // chunk-ext-name
-
-        // whole value part is optional. if no '=' expect next chunk-ext
-        if (ParseBws(tok) && tok.skip('=') && ParseBws(tok)) {
-
-            if (!skipKnown) {
-                if (ext.cmp("use-original-body",17) == 0 && tok.int64(useOriginBody, 10)) {
-                    debugs(94, 3, "Found chunk extension " << ext << "=" << useOriginBody);
-                    buf_ = tok.remaining(); // parse checkpoint
-                    continue;
-                }
-            }
+    while (parseOneChunkExtension(tok, skipKnown)) {
+        buf_ = tok.remaining(); // got one extension; there may be more
+    }
 
-            debugs(94, 5, "skipping unknown chunk extension " << ext);
+    if (skipLineTerminatorIfAny(tok)) {
+        buf_ = tok.remaining();
+        parsingStage_ = theChunkSize ? Http1::HTTP_PARSE_CHUNK : Http1::HTTP_PARSE_MIME;
+        return true; // reached the end of extensions (if any)
+    }
 
-            // unknown might have a value token or quoted-string
-            if (tok.quotedStringOrToken(value) && !tok.atEnd()) {
-                buf_ = tok.remaining(); // parse checkpoint
-                continue;
-            }
+    static const CharacterSet NonLF = (CharacterSet::LF).complement().rename("non-LF");
+    if (tok.skipAll(NonLF) && tok.skip('\n'))
+        throw TexcHere("cannot parse chunk extension"); // <garbage> CR*LF
+    return false; // need more data
+}
+
+bool
+Http::One::TeChunkedParser::parseOneChunkExtension(Tokenizer &tok, const bool skipKnown)
+{
+    if (!ParseBws(tok)) // Bug 4492: IBM_HTTP_Server sends SP after chunk-size
+        return false;
+
+    if (!tok.skip(';'))
+        return false;
+
+    if (!ParseBws(tok)) // Bug 4492: ICAP servers send SP before chunk-ext-name
+        return false;
+
+    SBuf ext;
+    if (!tok.prefix(ext, CharacterSet::TCHAR)) // chunk-ext-name
+        return false;
 
-            // otherwise need more data OR corrupt syntax
-            break;
+    if (ParseBws(tok) && tok.skip('=') && ParseBws(tok)) {
+        if (!skipKnown) {
+            if (ext.cmp("use-original-body", 17) == 0 && tok.int64(useOriginBody, 10)) {
```

Comment:
This code will return true on partial known extension such as a `use-original-body=123` prefix of the `use-original-body=12345` extension. AFAICT, this wrong return value will lead to bugs -- when the rest of the extension value (i.e., `45`) arrives, it will be misinterpreted as a malformed extension (not starting with `;`).

tok.int64() is very similar to tok.prefix(); its callers need the same careful treatment of partial values.